add(README.md): Documentation for the env-injector snapcraft extension (#1)
locnnil/env-inject-doc: README.md

Reviewer: I'd add a section on Using the extension and give a snapcraft yaml example with a couple of apps using the extension. Then move the usage snap option examples to there. Also, document the alias.

Author: Added a "**How to use it**" section, as on the others extensions.

# The env-injector extension

The env-injector extension allows developers to export snap options as environment variables within the snap sandbox.
These options can be configured using the [snap set command](https://snapcraft.io/docs/configuration-in-snaps) or via the [the snap API](https://snapcraft.io/docs/using-the-api).
This extension dynamically converts snap configuration options into environment variables, which are then accessible to the app within the snap sandbox.
For this to work, the snap options must follow a specific pattern that the extension can match.

> The extension also supports environment files, where the snap configuration options specify the path to the environment file.

Reviewer: The first sentence isn't accurate. I'd rewrite the whole paragraph with both functions in mind (snap options and env files):

```suggestion
The env-injector extension adds support for passing user-defined environment variables to snaps. 
By adding the extension to snapped apps, the user of the snap will be allowed to set environment variables via [snap configuration options](https://snapcraft.io/docs/configuration-in-snaps) or environment files and have them exported to the apps on startup.
```

## How to use it

Add `extensions: [ env-injector ]` to the application definition in your `snapcraft.yaml` file.
By default the extension defines an app alias in the format of an env variable to be used in the snap config with the same name as defined in your apps section of the `yaml`.

> But if you wish, you can avoid to use this default by creating an environment Variable yourself with named as `env_alias` with the name that you wish to use to set the environment variables to your app.

Reviewer: ```suggestion
```

Add `extensions: [ env-injector ]` to the application definition in your `snapcraft.yaml` file.

By default, this extension creates an environment variable with the same name as the app defined in the yaml, which can be used in the snap configuration.
However, if you prefer, you can define this environment variable yourself by using `env_alias` with the name you want to use for setting the environment variables in your app.

> - Default usage example:

Reviewer: ```suggestion
Example, adding the extension to an app named `rabbitmq`:
```

```yaml
apps:
  rabbitmq:
    command: bin/run.sh
    daemon: simple
    extensions: [ env-injector ]
```

> This way you can define environment variables to your application by doing:

Reviewer: ```suggestion
Then, the user of the snap can pass environment variables via snap options or env files.

For example, to export `RABBITMQ_DEFAULT_PASS=password` using the `snap set` command:
```

|        | Snap Configuration Option      | Environment Variable |
|--------|--------------------------------|----------------------|
| Global | `env.<key>=<value>`            | `<KEY>=<value>`      |
| Local  | `apps.<app>.env.<key>=<value>` | `<KEY>=<value>`      |

**Examples**

Assuming `my-snap` is the name of the snap that uses the env-injector extension:

* `sudo snap set <my-snap> env.default-port=8080` will be mapped to `DEFAULT_PORT=8080` and exported globally.

* `sudo snap set <my-snap> apps.server.env.endpoint="/v2/store"` will be mapped to `ENDPOINT="/v2/store"` and exported only to the `server` app.

For environment files, where `path` refers to the env file path:

|        | Snap Configuration Option   |
|--------|-----------------------------|
| Global | `envfile=<path>`            |
| Local  | `apps.<app>.envfile=<path>` |

**Examples**

* `sudo snap set <my-snap> envfile=$HOME/envfile.env` makes all the environment variables globally visible.

* `sudo snap set <my-snap> apps.server.envfile=$HOME/envfile.env` makes the environment variables visible only to the `server` app.

> > [!NOTE]

Reviewer: This is Github-specific syntax.

# The env-injector extension

The env-injector extension adds support for passing user-defined environment variables to snaps. 
By adding the extension to snapped apps, the user of the snap will be allowed to set environment variables via [snap configuration options](https://snapcraft.io/docs/configuration-in-snaps) or environment files and have them exported to the apps on startup.

## Using the extension

Add `extensions: [ env-injector ]` to the application definition in your `snapcraft.yaml` file.

Example, adding the extension to an app named `server`:
```yaml
apps:
  server:
    command: run.sh
    daemon: simple
    extensions: [ env-injector ]
```

> With this, the user of the snap will be enabled to pass environment variables via snap options or env files.

Reviewer 2: s/enabled/able

Author: Done! Thanks!

### Pass environment variables via snap configuration options

For example, to set `HTTP_PORT=8080` for `web` app, the user can run:

```bash
sudo snap set <snap-name> env.http-port=8080
```

Internally, the extension converts `http-port` to `HTTP_PORT` and exports it before executing the app's command.

The above syntax is to set environment variables for all the apps that use this extension within the snap.
We refer to these as *global options*. 

> With the given example, the snap only has one app which uses the extension so setting a global option is acceptable. 

Reviewer 2: `extension, `

If the snap has multiple apps that need support for environment variables with conflicting variable names, then the user can target the desired app by setting a *local option*.
To set an environment variable for a single app, the user needs to add an `apps.<app>` prefix to the snap config option.

Using the same example as above but targetting only the `server` app:
```bash
sudo snap set <snap-name> apps.server.env.http-port=8080
```

The app's name is taken from the definition in the snapcraft YAML file.
This can be overridden with an alias; see [env alias](#env-alias).

Refer [here](#syntax) for more details on the syntax.

### Pass environment variables via env files

A list of environment variables can be passed to the snap via one or more environment files.

> **Note**  

> > For confined snaps, the files need to be accessible from within the snap, either by placing them in the [writable area](https://snapcraft.io/docs/data-locations) of the snap or via a file access interface, such as the [home](https://snapcraft.io/docs/home-interface) or [personal-files](https://snapcraft.io/docs/personal-files-interface).

Reviewer 2: Either drop `the` before `[home]`, or add `interface` at the end.

For example, to export the content of `/var/snap/my-snap/common/server.env` to `server` app:
```bash
sudo snap set <my-snap> apps.server.envfile=/var/snap/my-snap/common/server.env
```

The app's name is taken from the definition in the snapcraft YAML file.
This can be overridden with an alias; see [env alias](#env-alias).

Refer [here](#syntax) for more details on the syntax.

## Syntax and Rules

To avoid conflicts with existing snap options, some rules are defined below, along with the allowed syntax accepted by the extension exporter program.

### Syntax

The snap configuration options expect the following syntax for the keys:

* **Global:** Environment variables are visible to all apps within the snap that use this extension.
* **Local:** Environment variables are visible only to the specific `app` defined during the snap set command.

Snap configuration options to environment variable mapping:

|        | Snap configuration option      | Environment variable |
|--------|--------------------------------|----------------------|
| Global | `env.<key>=<value>`            | `<KEY>=<value>`      |
| Local  | `apps.<app>.env.<key>=<value>` | `<KEY>=<value>`      |

> The `key` is converted to environment variable name based on the following rules:

Reviewer 2: `...to an environment..`

| Accepted characters in snap option keys                   | Mapped environment variable names |
|-----------------------------------------------------------|-----------------------------------|
| Lowercase letters                                         | Uppercase letters                 |

> | Numbers (not at the beginning and with lowercase letters) | Numbers                           |

Reviewer 2: I don't understand why `and with lowercase letters` is mentioned here

> | Hyphens (surrounded by lowercase letters)                 | Underscores                       |

Reviewer 2: I don't think it's necessary to add the phrase in brackets. It should be clear enough from the first line in this table.

> Any characters not listed in the first column of the above table are not allowed.

Reviewer 2: `the above table` => `the table above`

or just say:
`Any characters not listed in this table are not allowed.`

Snap configuration options for setting env file paths:

|        | Snap configuration option   |
|--------|-----------------------------|
| Global | `envfile=<path>`            |
| Local  | `apps.<app>.envfile=<path>` |

### Order precedence

The environment variables are processed in the following order:

> 1. Env vars extracted from the **global** env file.
> 2. Env vars extracted from the **local** env file.
> 3. Env vars from global snap config options.
> 4. Env vars from local snap config options.

Reviewer: ```suggestion
1. Env vars extracted from the global env file.
2. Env vars extracted from the local env file.
3. Env vars from global snap config options.
4. Env vars from local snap config options.
```

> # Environment Variables Injector

Reviewer: ```suggestion
# The env-injector extension
```

Consistent with other extension docs: https://snapcraft.io/docs/supported-extensions

Author: Yeah, that makes sense! Thanks!

> This extension is designed to help users export snap options configured with the [snap set command](https://snapcraft.io/docs/configuration-in-snaps) into environment variables.

Reviewer: There are a few ways to set configurations, not just via the `snap set` command.

Author: Addressed, thank you!

> Additionally, this extension supports environment files, which we refer to as `envfile` for brevity.

Reviewer: The introduction needs improvements, keeping the following points in mind:
- The extension is to inject environment variables into snaps
- The extension offers two ways to do it
  - Via snap configuration options
  - Via environment files or env files

Author: Addressed on https://github.com/locnnil/env-inject-doc/pull/1/commits/1f0181a11f9680f197444887ec495a6854d5e13b! Thanks!

To prevent conflicts with other snap configurations and to maintain some level of isolation between apps, the following rules apply:

* **Global:** Environment variables are visible to all apps within the snap that use this extension.
* **Local:** Environment variables are visible only to the specific `app` defined during the snap set command.

|        | Snap Configuration Option      | Environment Variable |
|--------|--------------------------------|----------------------|

> | Global | env.<key>=<value>              | <KEY>=<value>        |

Reviewer: Avoid HTML syntax as they don't get rendered as intended.
```suggestion
| Global | `env.<key>=<value>`              | `<KEY>=<value>`        |
```

Author: Fixed! Thanks!

| Local  | apps.<app>.env.<key>=<value>   | <KEY>=<value>        |

**Examples**

Assuming `my-snap` is the name of the snap that uses the env-injector extension:

> * `sudo snap set <my-snap> env.default_port=8080` will be mapped to `DEFAULT_PORT=8080` globally.

Reviewer: ```suggestion
* `sudo snap set <my-snap> env.default-port=8080` will be mapped to `DEFAULT_PORT=8080` and exported globally.
```

> * `sudo snap set <my-snap> apps.server.env.endpoint="/v2/store"` will be mapped to `ENDPOINT="/v2/store"` and will be visible only to the `server` app.

Reviewer: Turn into a subsection, followed by another one for environment files

Author: Done in https://github.com/locnnil/env-inject-doc/pull/1/commits/1f0181a11f9680f197444887ec495a6854d5e13b.

> * `sudo snap set <my-snap> apps.server.env.endpoint="/v2/store"` will be mapped to `ENDPOINT="/v2/store"` and will be visible only to the `server` app.

Reviewer: ```suggestion
* `sudo snap set <my-snap> apps.server.env.endpoint="/v2/store"` will be mapped to `ENDPOINT="/v2/store"` and exported only to the `server` app.
```

> For environment files, where `path` refers to the `envfile` path:

Reviewer: `envfile` is not an alias for environment files. It is part of the snap option key for environment file / env file path.

`envfile` - > env file

Author: Sure, fixed! On 033f468

|        | Snap Configuration Option    |
|--------|------------------------------|
| Global | envfile=<path>               |
| Local  | apps.<app>.envfile=<path>    |

**Examples**

* `sudo snap set <my-snap> envfile=$HOME/envfile.env` makes all the environment variables globally visible.

> * `sudo snap set <my-snap> apps.server.envfile=$HOME/envfile.env` makes the environment variables visible only to the `server` app.

Reviewer: File access requirements not explained.

Author: Explained, thanks!

## Syntax and Rules

The precedence of environment variables follows this order:

1. Env vars from local snap config options.
2. Env vars from global snap config options.
3. Env vars extracted from the **local** `envfile`.

> 4. Env vars extracted from the **global** `envfile`.

Reviewer: Reading this sounds like the global env file has the highest precedence, applied in the end.

Author: Actually, this is the order of precedence that we use, but we can also list by **the order which is processed/applied**, which will be exactly the inverse of this list.

Reviewer: You could change the first sentence to "The environment variables are processed in the following order:" to avoid ambiguity.

Author: Done! Thank you!

> 3. Env vars extracted from the **local** `envfile`.
> 4. Env vars extracted from the **global** `envfile`.

Reviewer: ```suggestion
3. Env vars extracted from the **local** env file.
4. Env vars extracted from the **global** env file.
```

| Accepted Characters in Snap Option Keys                   | Mapped Environment Variable Names |
|-----------------------------------------------------------|-----------------------------------|
| Lowercase letters                                         | Uppercase letters                 |
| Numbers (not at the beginning and with lowercase letters) | Numbers                           |
| Hyphens (surrounded by lowercase letters)                 | Underscores                       |

> Any characters not listed in the first column of the above table are not allowed.

Reviewer: Move to snap options subsection.

Author: Moved! Thank you!

## Exporter Program Workflow

The process works as follows:

### During Build:

> - An application responsible for dynamically exporting environment variables is added to the snap's [command chain](https://snapcraft.io/docs/snapcraft-app-and-service-metadata#heading--command-chain).

Reviewer: ```suggestion
- An application responsible for dynamically exporting environment variables is added to the needed apps's [command chain](https://snapcraft.io/docs/snapcraft-app-and-service-metadata#heading--command-chain).
```

### At Runtime:

- The exporter program runs before the snapped application.

- It makes a request to the [snapd API](https://snapcraft.io/docs/using-the-api) through the snapd Unix socket.

> - It reads the available environment variables or paths to the envfiles.
> 
> - It translates the snap options into environment variables.
> 
> - If available, it sources the envfiles.
> 
> - It sources all envfiles.

Reviewer: ```suggestion
- It reads the available environment variables or paths to the env files.

- It translates the snap options into environment variables.

- If available, it sources the env files.

- It sources all env files.
```

> - It sources all envfiles.

Reviewer: Overlapping logic.

Author: Fixed! Thank you!

> - It sources all variables.

Reviewer: Not sure what this means.

Author: First it processes the env files, then it processes the snap configs.

| Local  | `apps.<app>.envfile=<path>` |

| Accepted characters in snap option keys                   | Mapped environment variable names |
|-----------------------------------------------------------|-----------------------------------|
| Lowercase letters                                         | Uppercase letters                 |
| Numbers (not at the beginning and with lowercase letters) | Numbers                           |
| Hyphens (surrounded by lowercase letters)                 | Underscores                       |

Any characters not listed in the first column of the above table are not allowed.

> 

Reviewer: Two issues with https://github.com/locnnil/env-inject-doc/pull/1/commits/55a87ea47cc8f9db8d7720bb039c4f80b397b44a:
- Line 93 which introduces the table hasn't been moved.
- This move results in confusion. The character table is unrelated to env file options but rather there to explain the mapping between snap options and environment variable names.

Author: True! Fixed!  Thanks!
